File: docker-cam-wrapper/camera_publisher/camera_publisher/camera_pub.py
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Image
import cv2
from cv_bridge import CvBridge, CvBridgeError


class CameraPublisher(Node):
    def __init__(self):
        super().__init__('camera_publisher')

        self.declare_parameter('device_path', '/dev/video0')
        device_path = self.get_parameter('device_path').get_parameter_value().string_value

        self.bridge = CvBridge()
        self.cap = cv2.VideoCapture(device_path)
        self.cap.set(cv2.CAP_PROP_FPS, 30)
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)

        if not self.cap.isOpened():
            self.get_logger().error(f'Failed to open camera at {device_path}')
            raise RuntimeError(f'Cannot open camera: {device_path}')

        self.pub = self.create_publisher(Image, '/image_raw', qos_profile_sensor_data)
        self.timer = self.create_timer(1.0 / 30.0, self.timer_callback)
        self.get_logger().info(f'Camera opened at {device_path}')

    def timer_callback(self):
        ret, frame = self.cap.read()
        if not ret:
            self.get_logger().warn('Failed to capture frame')
            return
        try:
            msg = self.bridge.cv2_to_imgmsg(frame, 'bgr8')
            msg.header.stamp = self.get_clock().now().to_msg()
            self.pub.publish(msg)
        except CvBridgeError as e:
            self.get_logger().error(f'CvBridgeError: {e}')

    def destroy_node(self):
        if self.cap.isOpened():
            self.cap.release()
            self.get_logger().info('Camera released.')
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = None
    try:
        node = CameraPublisher()
        rclpy.spin(node)
    except RuntimeError as e:
        print(f'[ERROR] {e}')
    except KeyboardInterrupt:
        pass
    finally:
        if node:
            node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: docker-cam-wrapper/camera_publisher_cpp/src/camera_pub.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <chrono>
#include <stdexcept>

class CameraPublisher : public rclcpp::Node
{
public:
  CameraPublisher()
  : Node("camera_publisher")
  {
    this->declare_parameter<std::string>("device_path", "/dev/video0");
    const std::string device_path = this->get_parameter("device_path").as_string();

    cap_.open(device_path);
    cap_.set(cv::CAP_PROP_FPS, 30);
    cap_.set(cv::CAP_PROP_FRAME_WIDTH, 640);
    cap_.set(cv::CAP_PROP_FRAME_HEIGHT, 480);

    if (!cap_.isOpened()) {
      RCLCPP_ERROR(this->get_logger(), "Failed to open camera at %s", device_path.c_str());
      throw std::runtime_error("Cannot open camera: " + device_path);
    }

    pub_ = this->create_publisher<sensor_msgs::msg::Image>(
      "/image_raw", rclcpp::SensorDataQoS());

    timer_ = this->create_wall_timer(
      std::chrono::milliseconds(33),
      std::bind(&CameraPublisher::timer_callback, this));

    RCLCPP_INFO(this->get_logger(), "Camera opened at %s", device_path.c_str());
  }

  ~CameraPublisher()
  {
    if (cap_.isOpened()) {
      cap_.release();
      RCLCPP_INFO(this->get_logger(), "Camera released.");
    }
  }

private:
  void timer_callback()
  {
    cv::Mat frame;
    if (!cap_.read(frame) || frame.empty()) {
      RCLCPP_WARN(this->get_logger(), "Failed to capture frame");
      return;
    }
    auto msg = cv_bridge::CvImage(
      std_msgs::msg::Header(), "bgr8", frame).toImageMsg();
    msg->header.stamp = this->now();
    pub_->publish(*msg);
  }

  cv::VideoCapture cap_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  try {
    rclcpp::spin(std::make_shared<CameraPublisher>());
  } catch (const std::runtime_error & e) {
    RCLCPP_ERROR(rclcpp::get_logger("main"), "%s", e.what());
  }
  rclcpp::shutdown();
  return 0;
}
